toolchain-analysis/bazel.md

```diff
@@ -0,0 +1,41 @@
+# Bazel toolchain analysis
+
+[Bazel](https://bazel.build/) is an open-source build and test tool similar to
+Make, Maven, and Gradle. It uses a human-readable, high-level build language.
+Bazel supports projects in multiple languages and builds outputs for multiple
+platforms. Bazel supports large codebases across multiple repositories, and
+large numbers of users.
+
+## Supported platforms
+
+TODO: test Bazel against currently Node.js supported platforms and toolchains
+(see [BUILDING.MD](https://github.com/nodejs/node/edit/master/BUILDING.md)).
+
+### Unsupported platforms
+
+TODO: see above
+
+## Pros
+
+* Easier to upgrade V8 as we don't need to rewrite build configs for V8
+* Compared to GN, Bazel is more widely adopted outside Chromium
+
+## Cons
+
+* Other dependencies will need Node.js-maintained GN build files
+* Currently support a limited set of platforms
+* V8 team can't guarantee Bazel will still be maintained if V8 stops being used
```

Comment:
I think a pretty big con here is that bazel would require users to install Java (right?) in addition to having to install Python for backwards compatibility with GYP)

Reply:
Not according to a few comments on https://github.com/nodejs/TSC/issues/901, so unless someone can demonstrate otherwise I'm reluctant to add it as a Con

Reply:
Also I tried to avoid general Java-related cons here (makes contributing back to Bazel harder, for example) because it's been seen as FUD/irrational in the past. But no, from a personal perspective I'm not trilled about the idea of a Java-based build toolchain.

Reply:
Ah, didn't have full context I was just thinking of it as another thing to install like we have Python. I don't actually mind the fact it's Java as much as it's another heavy thing users will have to install in addition to Python.

Reply:
I think any solution will require users (to be clear, users=node core contributors, not all node users) to install something in addition to Python.

Reply:
@targos won't this solution require anyone who is building a native addon to install Java?

I think there is a difference between having a binary vs. having another runtime. It's kind of hard for me to articulate...

I've seen a lot of people (not node core contributors) struggle with Python paths, installing the right Python etc. I suspect Java will come with the same sort of issues (not because it's worse/better or anything just because it's one more thing)

Reply:
This is only about building Node.js itself, not native addons.

Reply:
https://docs.bazel.build/versions/5.0.0/install-compile-source.html#bootstrap-windows Does look like it does need JDK in some cases. I agree that the differentiation between core vs. addons is importat though.

Reply:
This seems to say that they ship their own bundled JRE, so JDK might be optional unless actually building Java projects https://docs.bazel.build/versions/5.0.0/install-ubuntu.html#step-3-install-a-jdk-optional

Reply:
That's correct, a minimal JDK for Bazel's server is bundled into the release, there will be no Java dependency for node users.